clustermesh: add MCS-API ServiceImport controller

MrFreezeex opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

This adds the last needed MCS-API controller to sync ServiceImport from remote clusters in the clustermesh. This PR also  comes with an example that will be later used in the user documentation, small fixes to the helm deployment so that this controller is able to patch statuses and a small typo fix with the interface that was added for this controller in previous commits.

Fixes: #32712
Related to #27902

```release-note
clustermesh: add Multi-cluster Service API support
```

MrFreezeex committed:
clustermesh: fix typo in argument of the ClusterMesh interface

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
clustermesh: add MCS-API ServiceImport controller

This adds the last needed MCS-API controller to sync ServiceImport from
remote clusters in the clustermesh.

This controller pulls ServiceExport info from remote clusters and merge
those into a ServiceImport that it creates. The conflict resolution
algorithm favors the oldest ServiceExport for every properties excepts
the service ports. Service ports are the union of every similar
ServiceExports from the local cluster and the remote clusters.

The Service Import controller also manages ServiceExport status as it
contains the conflict messages and we already have all of this info when
we run the merge to create the ServiceImport in the corresponding
controller.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex committed:
helm: add missing permission to update MCS API status resources

MCS-API Service{Import,Export} also have statuses that we have to update
for MCS-API implementation so this adds the missing permissions.

This commit also remove permission to MCS-API resources when only
endpointslice sync is enabled as this is not necessary.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

squeed commented on pkg/clustermesh/mcsapi/cell.go:
```diff
@@ -113,5 +115,27 @@ func initMCSAPIController(params mcsAPIParams) error {
+			return nil // The parent context expired, and we are already terminating
+		}
+
+		if err := svcImportReconciler.SetupWithManager(params.CtrlRuntimeManager); err != nil {
```
OneShot jobs should terminate when the supplied `ctx` is closed. I don't see `ctx` being passed to `SetupWithManager()`; unless there's something I'm missing, there's nothing that will cleanly stop this controller?

MrFreezeex replied:
Hmm I am not sure how controller-runtime controllers handle their shutdown TBH. From a quick look I haven't been able to find some other cilium controller-runtime controllers that do something differently from this :/. So maybe there is something magic like if the informers are stopped, controllers like that are also stopped as a result 🤷‍♂️?

FYI the one shot job is to make sure the `ServiceExportsSynced` func doesn't hold the start of other hive components in the operator.

squeed replied:
aha, I missed that this was a shared controller-runtime manager! In which case, disregard this :-)

squeed commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+			return controllerruntime.Fail(err)
+		}
+	} else {
+		if err := r.Client.Update(ctx, svcImport); err != nil {
```
should this use server-side apply with field managers?

MrFreezeex replied:
IIUC this is useful when there are multiple actors that can update a kubernetes object right? In this specific case cilium operator is the only actor (not even a human should be manually editing this) updating ServiceImport resources so I would say that it is not necessary I think. But correct me if I'm wrong happy to take a look at this!

giorio94 replied:
Another possibility could be using `controllerutil.CreateOrUpdate`, to unify the two code paths.

squeed replied:
Makes sense; if there's only one owner, then there's no need to use SSA. Just checking :-)

squeed commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+	return ""
+}
+
+func (r *mcsAPIServiceImportReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
Can you write a short blurb describing the primary inputs and outputs? What is the object for req? What "downstream" objects are affected? Thanks!

MrFreezeex replied:
I will definitely write some code comments here (and update this thread once done), thanks for the feedback! If you want the bigger picture though (and not only from this controller perspective) I recommend looking at the schema here: https://github.com/cilium/cilium/issues/32712.

squeed replied:
That's a great diagram, thanks! Translating that to a docblock somehwere would be very helpful, as the context will be lost in the future.

MrFreezeex replied:
I tried to expand a bit the explanation on `mcsAPIServiceImportReconciler`  doc and yes I could definitely persist some similar diagram/schema somewhere in the repo. Do you have a suggestion as to where that should be though?

squeed replied:
Probably the best way to preserve the document is as a mermaid diagram in a README.md in a reasonable folder.
(Sorry for missing your comment :bow:)

MrFreezeex replied:
Ah ok I see! I will do that in the mcsapi folder then in a follow-up PR then, thanks!

giorio94 commented on install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml:
```diff
@@ -324,6 +324,13 @@ rules:
 {{- end }}
 {{- end }}
 {{- if or .Values.clustermesh.enableEndpointSliceSynchronization .Values.clustermesh.enableMCSAPISupport }}
```
Slightly unrelated, but it seems that these permissions are not required if only epslice synchronization is enabled (and not MCS API support). It looks like that this change got accidentally introduced in https://github.com/cilium/cilium/pull/32972/commits/a7cd30fa2fdb5e69450bc4e77c389cfcc3c04b51, while updating the other usages of that variable.

MrFreezeex replied:
Yep indeed good catch, thanks :D

giorio94 commented on examples/kubernetes/clustermesh/mcs-api-example/cluster1.yaml:
```diff
@@ -0,0 +1,111 @@
+  name: rebel-base-mcs
+spec:
+  type: ClusterIP
+  clusterIP: None
```
Nit. Based on the service name, I'd expect this one to not be headless, and the one below to actually be headless.

giorio94 commented on examples/kubernetes/clustermesh/mcs-api-example/cluster1.yaml:
```diff
@@ -0,0 +1,111 @@
+    spec:
+      containers:
+      - name: rebel-base
+        image: docker.io/nginx:1.15.8
```
Nit. nginx 1.15.8 looks definitely old, but I guess that we have never defined a good strategy to keep example manifests up-to-date (and test that they still work). It is probably fine to keep them as is, considering that the same versions are also used in the already existing examples. 

One possibility could be to actually introduce here only the differences compared to the already existing examples, assuming that the rest is identical. That would remove a bit of duplication.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+	queue   workqueue.RateLimitingInterface
+}
+
+func (s *remoteClusterServiceSource) onClusterServiceExportUpdate(svcExport *mcsapitypes.MCSAPIServiceSpec) {
```
Nit. `onClusterServiceExportEvent`, as `Update` is a bit misleading considering that it also gets called in case of deletion events.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+		Complete(r)
+}
+
+// remoteClusterServiceSource is a source to watch remote cluste service.
```
Nit. `s/cluste/cluster/`

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+func (r *mcsAPIServiceImportReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.remoteClusterServiceSource.handler = &handler.EnqueueRequestForObject{}
+	return ctrl.NewControllerManagedBy(mgr).
+		Owns(&mcsapiv1alpha1.ServiceImport{}).
```
`Owns` looks redundant to me in this context, as the controller is already `For` ServiceImports.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+			}
+			key := portMergeKey{
+				Protocol:    port.Protocol,
+				AppProtocol: appProtocol,
```
Nit. `AppProtocol: ptr.Deref(port.AppProtocol, ""),`

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+			if port.AppProtocol != nil {
+				appProtocol = *port.AppProtocol
+			}
+			key := portMergeKey{
```
IIUC the KEP, the key would be the name of the port (as you cannot have two ports with the same name), rather than the port, protocol, app protocol triple: 

> Should a port name be used for multiple non-identical (port, protocol, appProtocol) service ports by different constituent services, the conflict resolution policy will determine which values are used by the derived service.

MrFreezeex replied:
So the port name logic checking was more on `checkConflictExportPorts` function so far BUT I did forget to check if we export a port that has the same name as another port indeed, will check how I should do that :+1: . So far I was only checking if we export the same port but with a different name, like catching that essentially:
```
    - name: http
      protocol: TCP
      port: 80
      targetPort: 8080
```
```
    - name: not-http
      protocol: TCP
      port: 80
      targetPort: 8080
```

giorio94 replied:
Yeah, AFAIU the two aspects that must not be duplicated are the [name](https://github.com/kubernetes/kubernetes/blob/4b8e819355d791d96b7e9d9efe4cbafae2311c88/pkg/apis/core/validation/validation.go#L5503-L5512), and the [(port, protocol) tuple ](https://github.com/kubernetes/kubernetes/blob/4b8e819355d791d96b7e9d9efe4cbafae2311c88/pkg/apis/core/validation/validation.go#L5446-L5457)(appProtocol does not seem to be significant instead). 

Indeed, both of the two following services get rejected by k8s:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: same-port-proto
  namespace: default
spec:
  ports:
  - name: foo
    port: 80
    protocol: TCP
    appProtocol: ssh
    targetPort: 9898
  - name: bar
    port: 80
    protocol: TCP
    appProtocol: http
    targetPort: 9898
  selector:
    app: podinfo
---
apiVersion: v1
kind: Service
metadata:
  name: same-name
  namespace: default
spec:
  ports:
  - name: foo
    port: 80
    protocol: TCP
    targetPort: 9898
  - name: foo
    port: 81
    protocol: TCP
    targetPort: 9898
  selector:
    app: podinfo
``` 

```
$ Service "same-port-proto" is invalid: spec.ports[1]: Duplicate value: core.ServicePort{Name:"", Protocol:"TCP", AppProtocol:(*string)(nil), Port:80, TargetPort:intstr.IntOrString{Type:0, IntVal:0, StrVal:""}, NodePort:0}
$ Service "same-name" is invalid: spec.ports[1].name: Duplicate value: "foo"
```

MrFreezeex replied:
Reading this again I now realize that I skipped the part where app protocol is not significant :sweat_smile:.

Having two similar ports (similar meaning the tuple of port + protocol) with different app protocol seems to be a byproduct of the check for two similar ports with the default service kubernetes code.

However in our case we have to try merging the different ports from different clusters together so we have the additional context that the app protocol is different on a similar port so I think we could raise the conflict message specifically towards this instead of checking simply for a duplicated port + protocol as it's more meaningful to users and it's also easier to implement in our case IMO.

I pushed a fixup commit as a proposal implementing that.

giorio94 reviewed (commented):
Sorry, I don't know why a bunch of comments got published earlier... In any case, just a few nits from an initial high level pass, but I still need to fully digest the logic and perform a few tests before providing more in-depth comments.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+		return
+	}
+	s.Logger.Infof("Queueing update from remote cluster %s/%s",
+		svcExport.Namespace, svcExport.Name)
```
Nit. This looks a debug level log. Additionally, the parameters should be provided as fields, rather than embedded as part of the message.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+			Name:      svcExport.Name,
+			Namespace: svcExport.Namespace,
+		},
+	}}, s.queue)
```
I wonder whether something like:

```go
s.queue.Add(ctrl.Request{NamespacedName: types.NamespacedName{
	Name:      svcExport.Name,
	Namespace: svcExport.Namespace,
}})
```

could be easier to read, reducing the number of indirections (and the need for the handler).

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+func (s *remoteClusterServiceSource) onClusterServiceExportUpdate(svcExport *mcsapitypes.MCSAPIServiceSpec) {
+	if s.ctx == nil || s.handler == nil || s.queue == nil {
+		return
+	}
```
I'd suggest adding a comment that here it is safe to skip the events received before starting the controller, as they would be anyhow generated by the enqueue function of the namespace watcher during the initial synchronization.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+		if k8sApiErrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
```
Nit. `return false, client.IgnoreNotFound(err)`

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+		if k8sApiErrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
```
Nit. `return nil, client.IgnoreNotFound(err)`

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+			return controllerruntime.Fail(err)
+		}
+		if localSvc == nil {
+			if meta.SetStatusCondition(&svcExport.Status.Conditions, metav1.Condition{
```
The ready and conflict conditions should be likely removed/set to false in this case (same below).

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+			readyStatus = metav1.ConditionTrue
+			changedCondition = meta.RemoveStatusCondition(&svcExport.Status.Conditions, mcsapiv1alpha1.ServiceExportConflict) || changedCondition
+		}
+		changedCondition = meta.SetStatusCondition(&svcExport.Status.Conditions, metav1.Condition{
```
The ready condition should be removed/set to false in case of conflicts.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+func getServiceImportStatus(svcExportByCluster operator.ServiceExportsByCluster) mcsapiv1alpha1.ServiceImportStatus {
+	clusters := []mcsapiv1alpha1.ClusterStatus{}
+	clusterNames := maps.Keys(svcExportByCluster)
+	sort.Strings(clusterNames)
```
Nit. `slices.Sort(clusterNames)`

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+			LastTransitionTime: metav1.Now(),
+		})
+
+		if oldestCluster != r.clusterName {
```
Based on the following wording from the KEP:

> If any properties have conflicting values that can not simply be merged, a ServiceExportConflict condition will be set on all ServiceExports for the conflicted service with a description of the conflict.

I'd check for conflicts and set the appropriate condition regardless of whether the local service export is actually the current source of truth or not. That would also simplify a bit the logic, and ensure that the conditions eventually match on all clusters (that have the given ServiceExport).

MrFreezeex replied:
I am not sure if I understand correctly what you mean here, if we are the local source of truth we can't really have a conflict afaict because we are the authority :thinking:. But I might be missing something though

giorio94 replied:
Hmm, in my view you'd still have a conflict, although it gets resolved in your favor. But you still have the underlying problem that the services are not aligned across all clusters, which may be worth outlining in the condition, so that it can be easily detected from any cluster (that has the service export). Does that make sense?

MrFreezeex replied:
Ah I see what you mean, yes that make sense and I think is also valuable for admins, so will look at doing this, thanks :eyes:!

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+		return fmt.Sprintf(
+			"Conflicting %s. %d/%d clusters disagree. Using \"%s\" from oldest service export in \"%s\".",
+			fieldName, conflictCount, clusterCount, oldestSvcField, oldestClusterName,
+		)
```
I wonder whether it could make sense to concatenate the messages in case multiple conflicts are found, to ensure that they are all displayed at the same time (rather than requiring to fix one to show the others).

MrFreezeex replied:
In theory yes that sounds like a good idea, in practice I am not sure how readable it would be. I haven't see yet a controller that was concatenating multiple errors in a condition somehow (but I am not sure if it's because of the complexity it would bring or because user wise it is not readable :/)

giorio94 replied:
Yeah, I'm also a bit doubtful on how the concatenated messages would be rendered. In any case, I'm perfectly fine with the current approach.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller_test.go:
```diff
@@ -0,0 +1,463 @@
+
+		svcImport, err := getServiceImport(c, key)
+		require.NoError(t, err)
+		require.NotNil(t, svcImport)
```
Nit. I'd check that the generated object matches the expected format.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller_test.go:
```diff
@@ -0,0 +1,463 @@
+	if err != nil {
+		if k8sApiErrors.IsNotFound(err) {
+			return nil, nil
+		}
```
Nit. I'd avoid filtering the NotFound errors, so that it can be asserted directly below (rather than checking that the object is not nil). At that point, you could just directly call `client.Get`, and get rid of these helpers.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller_test.go:
```diff
@@ -0,0 +1,463 @@
+		require.Contains(t, svcImport.Spec.Ports, mcsapiv1alpha1.ServicePort{
+			Name: "named",
+			Port: 80,
+		})
```
Nit. Could be replaced by `require.ElementsMatch`.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller_test.go:
```diff
@@ -0,0 +1,463 @@
+		})
+	}
+
+	for _, remoteFixture := range remoteSvcImportTestFixtures {
+		remoteFixture.ExportCreationTimestamp = newerTime
+	}
```
This snipped modifies global variables, and apparently breaks the test when run with count > 1:

```
$ go test ./pkg/clustermesh/mcsapi/ --run Test_mcsServiceImport_Reconcile --count 2 

--- FAIL: Test_mcsServiceImport_Reconcile (0.00s)
    --- FAIL: Test_mcsServiceImport_Reconcile/Conflict_test_conflict-type_with_remote_conflict (0.00s)
        serviceimport_controller_test.go:422: 
                Error Trace:    /home/marco/Documents/isovalent/cilium/pkg/clustermesh/mcsapi/serviceimport_controller_test.go:422
                Error:          Should be true
                Test:           Test_mcsServiceImport_Reconcile/Conflict_test_conflict-type_with_remote_conflict
...
```

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller_test.go:
```diff
@@ -0,0 +1,463 @@
+			require.Nil(t, meta.FindStatusCondition(svcExport.Status.Conditions, mcsapiv1alpha1.ServiceExportConflict))
+		})
+	}
+}
```
It would be nice to have a couple of additional tests to check that (a) a service import gets correctly deleted if no longer necessary, (b) a service import (and its status) get correctly updated following changes and (c) the conditions of service exports get correctly updated, e.g., once a conflict gets fixed, or a new one conflict arises, as all these cases are trickier to test E2E, and could regress otherwise.

giorio94 reviewed (changes requested):
A few more comments inline, but the overall logic looks good to me :rocket:. I'll perform a final test once addressed.

squeed commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,489 @@
+	return mcsapiv1alpha1.ServiceImportStatus{Clusters: clusters}
+}
+
+func checkConflictExport(svcExportByCluster operator.ServiceExportsByCluster, clusterName, oldestClusterName string) string {
```
nit: short docblock please :-)

giorio94 commented on install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml:
Nit. Could you please extend the commit description to also mention why the permissions are no longer added if `enableEndpointSliceSynchronization` is enabled, for future reference?

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,539 @@
+
+	s.Logger.
+		WithField(logfields.K8sNamespace, svcExport.Namespace).
+		WithField("ServiceExport", svcExport.Namespace).
```
Nit. `s/svcExport.Namespace/svcExport.Name`

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,542 @@
+					Namespace: req.Namespace,
+				},
+			}, false, nil
+		}
```
Should return the error if different from NotFound.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,542 @@
+		SessionAffinityConfig: svc.Spec.SessionAffinityConfig.DeepCopy(),
+	}
+	mcsAPISvcSpec.ExportCreationTimestamp = svcExport.CreationTimestamp
+	return mcsAPISvcSpec
```
Nit.

```suggestion
	return &mcsapitypes.MCSAPIServiceSpec{
		Name:                  svc.Name,
		Namespace:             svc.Namespace,
		Ports:                 ports,
		Type:                  mcsAPISvcType,
		SessionAffinity:       svc.Spec.SessionAffinity,
		SessionAffinityConfig: svc.Spec.SessionAffinityConfig.DeepCopy(),
		ExportCreationTimestamp: svcExport.CreationTimestamp
	}
```

giorio94 replied:
Additionally, the `Cluster` field should be probably initialized as well.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,542 @@
+			if meta.SetStatusCondition(&svcExport.Status.Conditions, metav1.Condition{
+				Type:               mcsapiv1alpha1.ServiceExportValid,
+				Status:             metav1.ConditionFalse,
+				LastTransitionTime: metav1.Now(),
```
Nit. The `LastTransitionTime` could be omitted, as automatically set by `SetStatusCondition`.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,542 @@
+}
+
+func (r *mcsAPIServiceImportReconciler) createOrUpdateServiceImport(ctx context.Context, desiredSvcImport *mcsapiv1alpha1.ServiceImport) (*mcsapiv1alpha1.ServiceImport, error) {
+	svcImport := desiredSvcImport.DeepCopy()
```
Nit. It is enough to construct an object with name and namespace initialized, without having to fully deepcopy it.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,542 @@
+		currVal = val2
+	}
+	hasLocalCluster := val1.clusterName == localClusterName || val2.clusterName == localClusterName
+	if existingConflict != "" || !hasLocalCluster || (val1.Name == val2.Name && ptr.Deref(val1.AppProtocol, "") == ptr.Deref(val2.AppProtocol, "")) {
```
AFAIU, the `hasLocalCluster` check should be dropped, as we want to emit conflict messages regardless of whether the local cluster is the source of truth or not.

MrFreezeex replied:
The `hasLocalCluster` was more meant to not raise conflict messages between two remote clusters (aka the local cluster is not present in this conflict) so like we do not raise conflict in this scenario for instance:
`local` -> no conflict with `oldest`
`oldest` -> conflict with `other`
`other` -> conflict with `oldest`

giorio94 replied:
Hmm, my consideration was based on the following wording from the KEP:

> Conflict resolution policy: If any properties have conflicting values that can not simply be merged, a ServiceExportConflict condition will be set on all ServiceExports for the conflicted service with a description of the conflict. The conflict will be resolved by assigning precedence based on each ServiceExport's creationTimestamp, from oldest to newest.

which suggests that the conflict condition should be set to all ServiceExports associated with that service. However, it is not totally clear whether that applies to ports conflict resolution as well, given that they are a bit special compared to the other properties.

MrFreezeex replied:
Arf indeed I will try to ask around because I am not sure this is super logical. Like for instance with the type of conflict message suggested in the KEP there is no way you would know that there is actually no problem with your local cluster as it only suggest the winning cluster (like in the example I gave above you would have a conflict message indicating that `oldest` win but not that the conflict was between `oldest` and `other` which is IMO confusing if you are looking it from the `local` cluster view).

MrFreezeex replied:
So I asked in #sig-multicluster on the kubernetes slack (https://kubernetes.slack.com/archives/C09R1PJR3/p1725449448075499) and Tom Pantelis, a Submariner maintainer, answered that they were not technically able to respect that on the Submariner project (apparently they currently don't have all the service exports info from all the clusters like we do) and that he would like that the KEP no longer requires the conflict message on all the service exports.

I don't think we would have more answers until the next sig-multicluster meeting which should be happening the 17/09.  I will try to ask again the same question there to see what should ideally be happening from my little 3 clusters example mentioned in this thread and if as a result we should update the KEP or not.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,542 @@
+	if val1.Name != val2.Name {
+		return currVal, fmt.Sprintf(
+			"Conflicting port name on \"%v\". Using the port name \"%s\" from oldest service export in cluster \"%s\".",
+			currVal, currVal.Name, currVal.clusterName,
```
This message is technically not 100% correct, as the comparison is only being performed between two clusters, and there's no guarantee that either is the oldest. It would be probably easier to split this in two steps, first computing the merged state (hence always picking the oldest), and then generating the conflict message.

Another potential problem is that with the current implementation the conflict message is not guaranteed to be stable across multiple runs if there are multiple conflicts, as it depends on maps ordering. However, this may lead to endless reconciliations, as updating the status triggers a new reconciliation, which in turn causes the status to be updated, and so on. `checkConflictExport` is affected by the same issue AFAICT.

MrFreezeex replied:
Ooh yes indeed I didn't think about this :/. Will try to rework the port conflict code indeed :+1:. As for the `checkConflictExport` yep indeed good catch, afaict it should be easy enough to fix that part fortunately.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,542 @@
+	}
+
+	hasLocalCluster := val1.clusterName == localClusterName || val2.clusterName == localClusterName
+	if existingConflict != "" || !hasLocalCluster || deleteKey == nil {
```
Same here about the `hasLocalCluster` check.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,542 @@
+	return currVal, fmt.Sprintf(
+		"Duplicated port name \"%s\". Using port definition \"%v\" from oldest service export in cluster \"%s\".",
+		currVal.Name, currVal, currVal.clusterName,
+	), deleteKey
```
Same here about the fact that the message may not be strictly correct, and is not guaranteed to be the same across multiple runs.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,542 @@
+	}
+	if key1 == key2 {
+		deleteKey = nil
+	}
```
This check could be moved right after the definition of the variables, returning early from the function.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,542 @@
+		return controllerruntime.Fail(err)
+	}
+
+	newStatus := getServiceImportStatus(svcExportByRemoteCluster)
```
Should this list include the local cluster as well, when a ServiceExport is present there?

MrFreezeex replied:
Hmm after re-reading the KEP it's not as clear as I thought. IMO let's add it back for now as including is more exhaustive/clearer for a cluster admin looking at the status I think. If there is a conformance test about that and it's checking something different from this I would try to see with sig-multicluster who needs to change this behavior :eyes:

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller_test.go:
```diff
@@ -0,0 +1,800 @@
+			NamespacedName: key,
+		})
+
+		require.NoError(t, err)
```
Could it make sense to additionally check that the ServiceImport hasn't been created (assuming that the fake client doesn't do any namespace checks)?

MrFreezeex replied:
Hmm yes I think you are right and we can do this check indeed

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller_test.go:
```diff
@@ -0,0 +1,800 @@
+		require.ElementsMatch(t, svcImport.Spec.Ports, []mcsapiv1alpha1.ServicePort{
+			{
+				Port: 8000,
+			}, {
+				Name: "named",
+				Port: 80,
+			},
```
I'd suggest having also a port which is present both locally and remotely (without conflicts), to check that merging works correctly also in that case.

giorio94 commented on examples/kubernetes/clustermesh/mcs-api-example/svcexports.yaml:
I wonder whether it could make sense to drop this example for the moment, and include it directly as part of the documentation in a follow-up PR. My main concern with standalone examples is that they lack the accompanying documentation and tend to get stale over time. WDYT?

giorio94 reviewed (changes requested):
Thanks! I've left a bunch of additional comments inline. The vast majority are nits, with just a couple more relevant concerning the handling of conflicts.

MrFreezeex committed:
clustermesh: add a name to MCSAPI service controller for name conflict

As this Service controller also are for ServiceImport (while it truly
isn't about ServiceImport as its just a trick documented in the code) it
conflict with the name of the actual ServiceImport controller. This
commit rename the controller to make sure it does not.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,546 @@
+		}
+	}
+
+	oldestCluster := getOldestRemoteClusterService(svcExportByCluster)
```
```suggestion
	oldestCluster := getOldestClusterService(svcExportByCluster)
```

Nit. The list of service exports may include the local one as well.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,546 @@
+		}
+		return 1
+	})
+	return orderedSvcExports
```
Nit.

```suggestion
	return slices.SortedFunc(maps.Values(svcExportByCluster), func(a, b *mcsapitypes.MCSAPIServiceSpec) int {
		if a.ExportCreationTimestamp.Equal(&b.ExportCreationTimestamp) {
			return strings.Compare(a.Cluster, b.Cluster)
		}
		if a.ExportCreationTimestamp.Before(&b.ExportCreationTimestamp) {
			return -1
		}
		return 1
	})
```

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,546 @@
+				}
+			}
+
+			portIndex := findSimilarPort(port, ports)
```
Nit. `slices.IndexFunc` may be a bit more idiomatic here.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,546 @@
+			}
+
+			foundPortByName := true
+			if port.Name != "" {
```
Any reason why this check is skipped if the port name is empty? I'd intuitively treat it the same of any other name, which should also allow to simplify the logic a bit. Related, I guess there should be an extra check to assert that if the empty port is present, then it is the only one, otherwise the service creation would be rejected.

MrFreezeex replied:
So this is a bit confusing because there is two logics that checks the port name. This part of the logic take two port that has the same name and tries to check if it is actually the same port (aka same port/proto). This means that in the case of an unnamed port we don't have to do this because we can have multiple unnamed port that are different which is a valid thing afaik.

And the other part below that case does the opposite, it takes two similar port and check if every properties match including port name (which would be this check https://github.com/MrFreezeex/cilium/blob/7d8f382952896570fa8529655688876712237808/pkg/clustermesh/mcsapi/serviceimport_controller.go#L196). And this AFAIU would trigger the conflict on the case you just presented (same port one unnamed and one named).

giorio94 replied:
> This means that in the case of an unnamed port we don't have to do this because we can have multiple unnamed port that are different which is a valid thing afaik.

The Kubernetes API Server actually rejects a service with multiple unnamed ports:

```bash
$ cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: foo
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
  - port: 90
    protocol: TCP
    targetPort: 90
  selector:
    app: foo
EOF

The Service "foo" is invalid: 
* spec.ports[0].name: Required value
* spec.ports[1].name: Required value
```

And the same also applies if there are multiple ports, and any is unnamed.

```bash
$ cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: foo
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
  - name: bar
    port: 90
    protocol: TCP
    targetPort: 90
  selector:
    app: foo
EOF

The Service "foo" is invalid: spec.ports[0].name: Required value
```

MrFreezeex replied:
Ahhh TIL indeed thanks, I will update it thanks!

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,546 @@
+
+			portIndex := findSimilarPort(port, ports)
+			if portIndex == -1 {
+				if !foundPortByName {
```
When could this be true? I'd expect this conflict case to have already been caught by the logic above, when checking for the duplicate port names (as the name does exist, but the port/proto do not match).

MrFreezeex replied:
I'm confused, did you commented the right line?

If yes the `foundPortByName` is here that we never override any port name that was previously set (so that we always keep the oldest one). And if it was about why this whole check exists I think I just answered in the previous thread as it's a bit related to your previous question as well.

giorio94 replied:
Yeah, my question was essentially when the `foundPortByName` check does apply, and I agree that it is related to the other thread. The only case I could think of was regarding the unnamed ports, but removing that one it should no longer be necessary AFAICT.

giorio94 reviewed (approved):
Hey @MrFreezeex,

Sorry for the delay here, I've been busy with a different project.

I reviewed the PR once more, and overall looks great to me :rocket:. I've just left a couple of non-blocking and clarification questions inline.

Just to clarify, in the last comment you said:

> TLDR: many discussions upstream and IMO let's just do what the KEP is saying right now and I will update Cilium progressively once the KEP moves.

Is the conclusion that the current KEP version implies that conflicts should be emitted only if the local service is involved? I'm fine with either approach, just confirming as my understanding was that all conflicts had to be reported (and you are essentially doing that for port conflicts).

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,535 @@
+				return isSimilarPort(port, currPort.ServicePort)
+			})
+			if portIndex == -1 {
+				if !foundPortByName {
```
This check should no longer be needed now, right? If the port name already exists either it is a conflict (and we already bailed out above), or the port/proto match, and then we are in the else case below.

MrFreezeex replied:
Ah that's what you meant above, sorry. Hmm yes indeed I think that's correct

giorio94 replied:
Yeah, sorry if I hadn't been clear. Just trying to keep the logic as simple and easy to read as possible :grin:

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,535 @@
+// the ServiceExport object. This function does not check for conflict on the
+// ports field this aspect should be done by mergePorts
+func checkConflictExport(svcExportByCluster operator.ServiceExportsByCluster, oldestCluster string) string {
+	orderedSvcExports := orderSvcExportByPriority(svcExportByCluster)
```
Nit. Could be computed by the caller just to save a bit of operations in case there are a lot of clusters. That would also make `getOldestClusterService` trivial, as it is always the first one once sorted.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,535 @@
+			name: "sessionAffinityConfig.clientIP", getterFunc: getSessionAffinityConfigFunc,
+		},
+	}
+	for i, svcExport := range orderedSvcExports {
```
```suggestion
	for i, svcExport := range orderedSvcExports[1:] {
```

To avoid checking the first, which is always the oldest.

giorio94 commented on pkg/clustermesh/mcsapi/serviceimport_controller.go:
```diff
@@ -0,0 +1,535 @@
+			oldestField := getterStruct.getterFunc(svcExportByCluster[oldestCluster])
+			if oldestField != getterStruct.getterFunc(svcExport) {
+				continue
+			}
```
This feels weird to me. I guess it should be:

```suggestion
			if oldestField == getterStruct.getterFunc(svcExport) {
				continue
			}
```

And then `conflictCount := 1` below.

giorio94 reviewed (approved):
Thanks!
